Compile fails with ffmpeg 5.0

CorsixTH fails to compile with FFmpeg 5.0 released on January 17th 2022 due to incompatible changes to the API.

```
distcc[44717] ERROR: compile /home/stephen/corsixth/CorsixTH/Src/th_movie.cpp on 192.168.1.42:3636/8,lzo failed
distcc[44717] (dcc_build_somewhere) Warning: remote compilation of '/home/stephen/corsixth/CorsixTH/Src/th_movie.cpp' failed, retrying locally
distcc[44717] Warning: failed to distribute /home/stephen/corsixth/CorsixTH/Src/th_movie.cpp to 192.168.1.42:3636/8,lzo, running locally instead
/home/stephen/corsixth/CorsixTH/Src/th_movie.cpp: In member function ‘bool movie_player::load(const char*)’:
/home/stephen/corsixth/CorsixTH/Src/th_movie.cpp:343:52: error: invalid conversion from ‘AVCodec**’ to ‘const AVCodec**’ [-fpermissive]
  343 |                                            -1, -1, &video_decoder, 0);
      |                                                    ^~~~~~~~~~~~~~
      |                                                    |
      |                                                    AVCodec**
In file included from /home/stephen/corsixth/CorsixTH/Src/th_movie.h:47,
                 from /home/stephen/corsixth/CorsixTH/Src/th_movie.cpp:23:
/usr/include/libavformat/avformat.h:2165:41: note:   initializing argument 5 of ‘int av_find_best_stream(AVFormatContext*, AVMediaType, int, int, const AVCodec**, int)’
 2165 |                         const AVCodec **decoder_ret,
      |                         ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/home/stephen/corsixth/CorsixTH/Src/th_movie.cpp:355:52: error: invalid conversion from ‘AVCodec**’ to ‘const AVCodec**’ [-fpermissive]
  355 |                                            -1, -1, &audio_decoder, 0);
      |                                                    ^~~~~~~~~~~~~~
      |                                                    |
      |                                                    AVCodec**
In file included from /home/stephen/corsixth/CorsixTH/Src/th_movie.h:47,
                 from /home/stephen/corsixth/CorsixTH/Src/th_movie.cpp:23:
/usr/include/libavformat/avformat.h:2165:41: note:   initializing argument 5 of ‘int av_find_best_stream(AVFormatContext*, AVMediaType, int, int, const AVCodec**, int)’
 2165 |                         const AVCodec **decoder_ret,
      |                         ~~~~~~~~~~~~~~~~^~~~~~~~~~~
distcc[44721] ERROR: compile (null) on localhost failed
distcc[44717] ERROR: compile /home/stephen/corsixth/CorsixTH/Src/th_movie.cpp on localhost failed
make[2]: *** [CorsixTH/CMakeFiles/CorsixTH_lib.dir/build.make:426: CorsixTH/CMakeFiles/CorsixTH_lib.dir/Src/th_movie.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:360: CorsixTH/CMakeFiles/CorsixTH_lib.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

--- a/CorsixTH/Src/th_movie.cpp
+++ b/CorsixTH/Src/th_movie.cpp
@@ -338,7 +338,7 @@ bool movie_player::load(const char* szFilepath) {
     return false;
   }
 
-  AVCodec* video_decoder;  // unowned, do not free
+  av_codec_ptr video_decoder;  // unowned, do not free
   video_stream_index = av_find_best_stream(format_context, AVMEDIA_TYPE_VIDEO,
                                            -1, -1, &video_decoder, 0);
   if (video_stream_index < 0) {
@@ -350,7 +350,7 @@ bool movie_player::load(const char* szFilepath) {
       video_decoder, format_context->streams[video_stream_index]);
   avcodec_open2(video_codec_context.get(), video_decoder, nullptr);
 
-  AVCodec* audio_decoder;  // unowned, do not free
+  av_codec_ptr audio_decoder;  // unowned, do not free
   audio_stream_index = av_find_best_stream(format_context, AVMEDIA_TYPE_AUDIO,
                                            -1, -1, &audio_decoder, 0);
   if (audio_stream_index >= 0) {
@@ -363,7 +363,7 @@ bool movie_player::load(const char* szFilepath) {
 }
 
 av_codec_context_unique_ptr movie_player::get_codec_context_for_stream(
-    AVCodec* codec, AVStream* stream) const {
+    av_codec_ptr codec, AVStream* stream) const {
   av_codec_context_unique_ptr ctx(avcodec_alloc_context3(codec));
   avcodec_parameters_to_context(ctx.get(), stream->codecpar);
   return ctx;

--- a/CorsixTH/Src/th_movie.h
+++ b/CorsixTH/Src/th_movie.h
@@ -50,6 +50,12 @@ extern "C" {
 #include <libswscale/swscale.h>
 }
 
+#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(59, 0, 100)
+using av_codec_ptr = AVCodec*;
+#else
+using av_codec_ptr = const AVCodec*;
+#endif
+
 //! \brief Functor for deleting AVPackets
 //!
 //! Deletes AVPacket pointers that are allocated with av_malloc
@@ -359,7 +365,7 @@ class movie_player {
 
   //! Get the AVCodecContext associated with a given stream
   av_codec_context_unique_ptr get_codec_context_for_stream(
-      AVCodec* codec, AVStream* stream) const;
+      av_codec_ptr codec, AVStream* stream) const;
 
   //! Get the time the given frame should be played (from the start of the
   //! stream)

